pkg/reconciler/finalizer/reposync_finalizer.go

```diff
@@ -0,0 +1,220 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package finalizer
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"k8s.io/klog/v2"
+	"kpt.dev/configsync/pkg/api/configsync/v1beta1"
+	"kpt.dev/configsync/pkg/applier"
+	"kpt.dev/configsync/pkg/reposync"
+	"kpt.dev/configsync/pkg/status"
+	"kpt.dev/configsync/pkg/util/mutate"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// RepoSyncFinalizer handles finalizing RepoSync objects, using the Destroyer
+// to destroy all managed user objects previously applied from source.
+// Impliments the Finalizer interface.
+type RepoSyncFinalizer struct {
```

Comment:
I'm trying to find differences between `RepoSyncFinalizer` and `RootSyncFinalizer` but am struggling to find any. Are there any meaningful differences? I'm wondering if we can merge these into a generic finalizer struct.

Reply:
I built a generic one first and ended up with a train wreck of cascading type conversion switch statements. So I simplified it down to be a copy/paste/find/replace.

Reply:
IMO, the only real fix requires using generics, which I don't want to do yet.